src/physics/material/MaterialParams.hh

```diff
@@ -0,0 +1,99 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file MaterialParams.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "base/DeviceVector.hh"
+#include "base/Types.hh"
+#include "physics/base/Units.hh"
+#include "ElementDef.hh"
+#include "MaterialDef.hh"
+#include "Types.hh"
+#include "MaterialParamsPointers.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Data management for material, element, and nuclide properties.
+ */
+class MaterialParams
+{
+  public:
+    //@{
+    //! Type aliases
+    //@}
+
+    struct ElementInput
+    {
+        int            atomic_number; //!< Z number
+        units::AmuMass atomic_mass;   //!< Isotope-weighted average atomic mass
+        std::string    name;          //!< Element name
+    };
+
+    struct MaterialInput
+    {
+        real_type   number_density; //!< Atomic number density [1/cm^3]
```

Comment:
Yeah, that's definitely a more natural way of doing it. Right now though all the input is being preprocessed through Geant4, and number densities are a more physics-friendly system.

Reply:
We can actually add that to `MaterialInput`. The `GeantMaterial` struct provides both at export time.

Reply:
That would over-constrain the input though: density, number density, and elemental atomic masses are all related. Computationally we'll be using the number density, not the density, so it's most appropriate to preserve that number numerically rather than going back and forth from the elemental densities. Trust me on this one after having dealt with the SCALE/Shift/MCNP compositions/atomic numbers/nuclide metadata/physics mix tables. Users wonder why their exact number densities input values lose several digits of precision and it's the recalculation (and inconsistent masses/densities) that is the cause.

Reply:
Interesting. As far as I remember, Geant does the opposite: You provide the density (g/cm3) when constructing a new material, and Geant calculates the rest (atomic density and other derived quantities such as radiation length).

I also have a feeling that keeping the atomic density as a numerical value seems more appropriate for internal tasks, but I am not sure, i.e. I know models may call both atomic and macroscopic densities during their calculations, but I don't have a number to state which is more common. If we end up having a lot of `get_density()` calls all over the place, then it's better to cache that.

Reply:
Oh yeah we do cache the density for sure. I guess it's a "six of one, half a dozen of the other" situation. Is a material in Geant4 defined by *weight* fractions of the constituents, or by *number* fractions, or is it an option?

Reply:
It looks like either mass fraction or number of atoms (but not atom fraction).

Reply:
OK cool -- I guess "number of atoms" is just an unnormalized atom fraction.